File: N-Queens.java
class Solution {
    public boolean isSafe(int row, int col, char[][] board){
        for(int j = 0; j < board.length; j++){
            if(board[row][j] == 'Q'){
                return false;
            }
        }
        for(int i = 0; i < board.length; i++){
            if(board[i][col] == 'Q'){
                return false;
            }
        }
        for(int r = row, c = col; r >= 0 && c >= 0; r--, c--){
            if(board[r][c] == 'Q'){
                return false;
            }
        }
        for(int r = row, c = col; r >= 0 && c < board.length; r--, c++){
            if(board[r][c] == 'Q'){
                return false;
            }
        }
        for(int r = row, c = col; r < board.length && c >= 0; r++, c--){
            if(board[r][c] == 'Q'){
                return false;
            }
        }
        for(int r = row, c = col; r < board.length && c < board.length; r++, c++){
            if(board[r][c] == 'Q'){
                return false;
            }
        }
        return true;
    }

    public void saveBoard(char[][] board, List<List<String>> allBoards){
        String row = "";
        List<String> newBoard = new ArrayList();

        for(int i = 0; i < board.length; i++){
            row = "";
            for(int j = 0; j < board[0].length; j++){
                if(board[i][j] == 'Q')
                    row += 'Q';
                else 
                    row += '.';
            }
            newBoard.add(row);
        }
        allBoards.add(newBoard);
    }


    public void helper(char[][] board, List<List<String>> allBoards, int col){
        if(col == board.length){
            saveBoard(board, allBoards);
            return;
        }
        for(int row = 0; row < board.length; row++){
            if(isSafe(row, col, board)){
                board[row][col] = 'Q';
                helper(board, allBoards, col+1);
                board[row][col] = '.';
            }
        }
    }


    public List<List<String>> solveQueens(int n){
        List<List<String>> allBoards = new ArrayList();
        char[][] board = new char[n][n];
        helper(board, allBoards, 0);
        return allBoards;
    }
}

File: N_Queens.cpp
#include<iostream>
#include<vector>
#include<string>
using namespace std;

class Solution {
public:
    vector<vector<string>> all_ans;

    vector<vector<string>> solveNQueens(int n) {
        vector<string> board(n, string(n, '.'));
        solve(0, n, board);
        return all_ans;
    }

    bool isSafe(int row, int col, vector<string> &board){
        int n = board.size();
        for(int r = 0; r < n; r++){
            if(board[r][col] == 'Q')
                return false;
        }
        for(int c = 0; c < n; c++){
            if(board[row][c] == 'Q')
                return false;
        }
        for(int r = row, c = col; r >= 0 && c >= 0; r--, c--){
            if(board[r][c] == 'Q')
                return false;
        }
        for(int r = row, c = col; r < n && c < n; r++, c++){
            if(board[r][c] == 'Q')
                return false;
        }
        for(int r = row, c = col; r >= 0 && c < n; r--, c++){
            if(board[r][c] == 'Q')
                return false;
        }
        for(int r = row, c = col; r < n && c >= 0; r++, c--){
            if(board[r][c] == 'Q')
                return false;
        }
        return true;
    }

    void solve(int col, int &n, vector<string> &board){
        if(col == n){
            all_ans.push_back(board);
            return;
        }
        for(int row = 0; row < n; row++){
            if(isSafe(row, col, board)){
                board[row][col] = 'Q'; 
                solve(col + 1, n, board);
                board[row][col] = '.';
            }
        }
    }
};

int main(){
	int n;
	cin >> n;
	Solution sol;
	vector<vector<string>> answer = sol.solveNQueens(n);
	int sol_no = 0;
	for(auto a_v: answer){
		sol_no++;
		cout << sol_no<< "st sol --> ";
		for(auto a: a_v){
			cout << a << ", ";
		}	
		cout << endl;
	}
	return 0;
}
